### ISSUE TITLE
Code duplication in compiled client javascript when using include/extend: pug 2

### ISSUE THREAD

**Issue by noahmiller** [pugjs/pug#3103]
There was discussion in #1382 about how `include` and `extend` duplicate the code that is being included / extended.  That discussion ended with this note from @ForbesLindesay:

> Jade 2.0.0 will have some kind of builtin support for this via import for mixins. It will also allow providing custom linkers and loaders, which should make this possible to do in user space. Please open a new issue after 2.0.0 has been released if you want any help getting that working.

Does pug offer any of this support for reducing duplication?

For context, we have a few hundred lines of reusable form code included in a couple hundred form templates, and that has resulted into several hundred thousand duplicate lines in the final bundle.  We're looking for a way to reduce eliminate duplication.

Thanks!

**Comment by ForbesLindesay** [comment]
I'd say it's still early days, but it sort of provides the building blocks for what you're after. What would be needed is a custom implementation of the [PugLinker](https://github.com/pugjs/pug/tree/master/packages/pug-linker) that would emit shared files separately, rather than linking them inline. It's not a small task to build, and it's not top of my priority list personally, but it is now possible

**Comment by SerayaEryn** [comment]
I did some investigation how this could be implemented for mixins. Mixins are a bit easier since it is known which variables need to be passed to them.

I have a rough implementation that passes all tests cases (beside one where two mixins with the same name are being defined).

Maybe it is possible to build something for includes based on my implementation for mixins. 

I think I am going to open a PR once #3019 is merged.

**Comment by noahmiller** [comment]
@ForbesLindesay and @SerayaEryn, thanks for the heads up.

@SerayaEryn, your code to handle this for mixins assumes the mixins are in the same file as wherever they are being called?  We have a file of common mixins (and nothing else) that we include in other templates for certain common, reusable functionality.

**Comment by SerayaEryn** [comment]
My code stores any compiled mixin at the `pug-runtime` (which means it only works with inlineRuntimeFunctions=false).
Once a mixin has been stored at the `pug-runtime` it can be used in any template.

**Comment by noahmiller** [comment]
We currently compile each template to a stand-alone function prior to deployment, so it sounds like that wouldn't be directly applicable.  But I'd guess reviewing your code, once it's ready, would be very helpful towards implementing our goal.

**Comment by berndartmueller** [comment]
Are there any updates on this issue?
